proposals/0271-swiftpm-binary-dependencies.md

```diff
@@ -0,0 +1,417 @@
+* Ease of maintenance in the face of an evolving SwiftPM
+* Understandable composition with current and upcoming SwiftPM features
+* Support existing well-known occurrences of binary artifacts in the existing
+  (often iOS focused) target developer market.
+
+while keeping the following as non-goals:
+
+* Ease of production of binary packages
+* Simplicity of binary artifact distribution mechanism
+* Widespread use of binary packages
+
+## New `PackageDescription` API
+
+### BinaryTarget
+Since, a binary target is different compared to a source only target, we propose to introduce a new struct `Artifact`. This struct defines a targets associated artifacts.
+
+We propose to support local and remote artifacts from the beginning. In the alternatives considered section is larger collection of potential artifact stores. However we opted to simplify the initial implementation by just supporting a url and a path based definition. Later, we can implement different types of providers with different authentication methods.
+
+```swift
+public struct Artifact {
+    public enum Source {
+        case url(String, checksum: String)
+        case path
+    }
+
+    public let source: Source
+}
+```
+
+Furthermore, we propose to add a new `artifacts: [Artifacts]?` property to the `Target`, as well as extend the initializer with this parameter and create a new static method called `.binaryTarget()`. Lastly, we propose to extend the `TargetType` enum with a new case called `binary`.
+
+### PackageDescription
+To opt out of binary packages we propose a new configuration point inside the package description.
+
+```swift
+public final class Package {
+    ...
+    /// This disallows any binary dependency or any transitive binary dependency.
+    public var disallowsBinaryDependencies: Bool
+    ...
+}
+```
+
+## New `Package.resolved` Behavior
+
+For binary targets we store the checksum of the artifact in the `Package.resolved`. This lets us check for errors during resolution where a package's version did not change but the checksum did. In this case we will throw an error alerting the user about this.
+
+### Resolution
+
+Package resolution and dependency expression will not be impacted by this change (except where explicitly noted).
+
+#### Multiple references to same artifact
+During resolution SwiftPM will check that all references to an artifact in a dependency graph have the same checksum.
+
+## Binary Target Artifact Format
+
+SwiftPM currently supports multiple platforms; however, this proposal only proposes to add support for binary targets on Apple platforms. The reason for this is that Apple platform provide ABI guarantees and an already existing format we can leverage to make simplify the initial implementation. For Apple platforms we propose to use the `XCFramework` format for artifacts. This format already supports dynamic and static linking. Furthermore, it can contain products for every individual Apple platform at once.
```

Comment:
```suggestion
SwiftPM currently supports multiple platforms; however, this proposal only adds support for binary targets on Apple platforms. The reason for this is that Apple platforms provide ABI guarantees and an already existing format we can leverage to simplify the initial implementation. For Apple platforms we propose to use the `XCFramework` format for artifacts. This format already supports dynamic and static linking. Furthermore, it can contain products for every individual Apple platform at once.
```